File: src/main/java/org/goafabric/personservice/logic/PersonLogic.java
package org.goafabric.personservice.logic;

import io.quarkus.panache.common.Page;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;
import org.eclipse.microprofile.rest.client.inject.RestClient;
import org.goafabric.personservice.adapter.CalleeServiceAdapter;
import org.goafabric.personservice.controller.dto.Person;
import org.goafabric.personservice.controller.dto.PersonSearch;
import org.goafabric.personservice.persistence.PersonRepositoryPanache;

import java.util.List;

@ApplicationScoped
@Transactional
public class PersonLogic {
    private final PersonMapper personMapper;

    private final PersonRepositoryPanache personRepository;

    private final CalleeServiceAdapter calleeServiceAdapter;

    public PersonLogic(PersonMapper personMapper, PersonRepositoryPanache personRepository, @RestClient CalleeServiceAdapter calleeServiceAdapter) {
        this.personMapper = personMapper;
        this.personRepository = personRepository;
        this.calleeServiceAdapter = calleeServiceAdapter;
    }

    public Person getById(String id) {
        return personMapper.map(
                personRepository.findById(id));
    }


    public Person save(Person person) {
        return personMapper.map(
                personRepository.save(personMapper.map(person)));
    }

    public void delete(String id) {
        personRepository.deleteById(id);
    }

    public Person sayMyName(String name) {
        return new Person(null, null,
                calleeServiceAdapter.sayMyName(name).message(), "", null);
    }
    
    public List<Person> search(PersonSearch personSearch, Integer page, Integer size) {
        return personMapper.map(
                personRepository.find(personSearch, Page.of(page, size))
        );
    }

    public List<Person> findByStreet(String street, Integer page, Integer size) {
        return personMapper.map(
                personRepository.findByStreet(street, Page.of(page, size))
        );
    }
}

File: src/main/kotlin/org/goafabric/personservice/logic/PersonLogic.kt
package org.goafabric.personservice.logic

import jakarta.data.page.PageRequest
import jakarta.enterprise.context.ApplicationScoped
import jakarta.transaction.Transactional
import org.eclipse.microprofile.rest.client.inject.RestClient
import org.goafabric.personservice.adapter.CalleeServiceAdapter
import org.goafabric.personservice.controller.dto.Person
import org.goafabric.personservice.controller.dto.PersonSearch
import org.goafabric.personservice.extensions.UserContext
import org.goafabric.personservice.logic.mapper.PersonMapper
import org.goafabric.personservice.persistence.PersonRepository

@Transactional
@ApplicationScoped
class PersonLogic(
    private val personMapper: PersonMapper,
    private val personRepository: PersonRepository,
    @param:RestClient private val calleeServiceAdapter: CalleeServiceAdapter) {

    fun getById(id: String): Person {
        return personMapper.map(personRepository.findById(id))
    }

    fun search(personSearch: PersonSearch, page: Int, size: Int): List<Person> {
        val persons = personRepository.search(personSearch.firstName, personSearch.lastName,
            PageRequest.ofPage(page.toLong() + 1, size, true))
        return personMapper.map(persons)
    }

    fun save(person: Person): Person {
        return personMapper.map(
            personRepository.save(
                personMapper.map(person)))
    }

    fun delete(id: String) {
        personRepository.deleteById(id)
    }

    fun sayMyName (name : String) : Person {
        return Person(firstName = calleeServiceAdapter.sayMyName(name).message, lastName = "", address = emptyList())
    }

}